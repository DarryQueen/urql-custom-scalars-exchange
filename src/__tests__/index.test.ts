import {
  ExchangeIO,
  Operation,
  OperationResult,
  createClient,
} from '@urql/core';
import { DocumentNode, IntrospectionQuery } from 'graphql';
import gql from 'graphql-tag';
import { pipe, map, makeSubject, publish, tap } from 'wonka';

import scalarExchange from '../';
import schema from './__fixtures__/schema.json';

interface TestCase {
  query: DocumentNode;
  variables?: {};
  data: {};
  calls: number;
}

const dispatchDebug = jest.fn();

let client = createClient({ url: 'http://0.0.0.0' });
let { source: ops$, next } = makeSubject<Operation>();

beforeEach(() => {
  client = createClient({ url: 'http://0.0.0.0' });
  ({ source: ops$, next } = makeSubject<Operation>());
});

const simpleData = 'a';
const nestedData = { name: 'a' };

const simple: TestCase = {
  query: gql`
    {
      simple
    }
  `,
  data: { simple: simpleData },
  calls: 1,
};

const nested: TestCase = {
  query: gql`
    {
      nested {
        name
      }
    }
  `,
  data: { nested: nestedData },
  calls: 1,
};

const nestedNullable: TestCase = {
  query: gql`
    {
      nestedNullable {
        name
      }
    }
  `,
  data: { nestedNullable: null },
  calls: 0,
};

const list: TestCase = {
  query: gql`
    {
      list
    }
  `,
  data: { list: [simpleData, simpleData] },
  calls: 2,
};

const listNested: TestCase = {
  query: gql`
    {
      listNested {
        name
      }
    }
  `,
  data: { listNested: [nestedData, nestedData] },
  calls: 2,
};

const listNestedNullable: TestCase = {
  query: gql`
    {
      listNestedNullable {
        name
      }
    }
  `,
  data: { listNestedNullable: null },
  calls: 0,
};

const listNestedWithInput: TestCase = {
  query: gql`
    {
      listNested(input: $input) {
        name
      }
    }
  `,
  variables: { input: { topInput: 'topInput' } },
  data: { listNested: [nestedData, nestedData] },
  calls: 2,
};

const fragment1: TestCase = {
  query: gql`
    {
      ...QueryFields
    }

    fragment QueryFields on Query {
      listNested {
        name
      }
    }
  `,
  data: { listNested: [nestedData] },
  calls: 1,
};

const fragment2: TestCase = {
  query: gql`
    {
      listNested {
        ...ListFields
      }
    }

    fragment ListFields on Nested {
      name
    }
  `,
  data: { listNested: [nestedData, nestedData] },
  calls: 2,
};

const repeatedFragment: TestCase = {
  query: gql`
    fragment SomeFragment on Nested {
      name
    }

    {
      first: nested {
        ...SomeFragment
      }

      second: nested {
        ...SomeFragment
      }
    }
  `,
  data: { first: nestedData, second: nestedData },
  calls: 2,
};

const nestedFragment: TestCase = {
  query: gql`
    query {
      listNested {
        ...nested1
      }
    }
    fragment nested1 on Nested {
      name
      deeplyNested {
        ...nested2
      }
    }
    fragment nested2 on Nested {
      name
    }
  `,
  data: {
    listNested: [
      {
        name: 'firstLevel',
        deeplyNested: {
          name: 'secondLevel',
        },
      },
    ],
  },
  calls: 2,
};

const TEST_CASES: TestCase[] = [
  fragment1,
  fragment2,
  list,
  listNested,
  listNestedNullable,
  listNestedWithInput,
  nested,
  nestedNullable,
  repeatedFragment,
  simple,
  nestedFragment,
];

test.each(TEST_CASES)(
  'works on different structures',
  ({ query, variables, data, calls }) => {
    const op = client.createRequestOperation('query', {
      key: 1,
      query,
      variables,
    });

    const response = jest.fn(
      (forwardOp: Operation): OperationResult => {
        expect(forwardOp.key === op.key).toBeTruthy();
        expect(forwardOp.variables).toMatchSnapshot('Variables');
        return {
          operation: forwardOp,
          data: { __typename: 'Query', ...data },
        };
      }
    );
    const result = jest.fn();
    const forward: ExchangeIO = ops$ => pipe(ops$, map(response));

    const scalars = {
      String: jest.fn((text: string) => {
        return text.toUpperCase();
      }),
    };

    pipe(
      scalarExchange({
        schema: (schema as unknown) as IntrospectionQuery,
        scalars,
      })({
        forward,
        client,
        dispatchDebug,
      })(ops$),
      map(operationResult => {
        expect(operationResult.data).toMatchSnapshot('Output');
        return operationResult;
      }),
      tap(result),
      publish
    );

    next(op);

    expect(scalars.String).toHaveBeenCalledTimes(calls);
    expect(result).toHaveBeenCalledTimes(1);
  }
);
