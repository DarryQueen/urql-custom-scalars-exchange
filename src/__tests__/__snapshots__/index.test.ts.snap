// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works on the fragment1 structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNested": Array [
    Object {
      "name": "A",
    },
  ],
}
`;

exports[`works on the fragment2 structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNested": Array [
    Object {
      "name": "A",
    },
    Object {
      "name": "A",
    },
  ],
}
`;

exports[`works on the list structure: Output 1`] = `
Object {
  "__typename": "Query",
  "list": Array [
    "A",
    "A",
  ],
}
`;

exports[`works on the listNested structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNested": Array [
    Object {
      "name": "A",
    },
    Object {
      "name": "A",
    },
  ],
}
`;

exports[`works on the listNestedNullable structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNestedNullable": null,
}
`;

exports[`works on the listNestedWithInput structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNested": Array [
    Object {
      "name": "A",
    },
    Object {
      "name": "A",
    },
  ],
}
`;

exports[`works on the listNestedWithInput structure: Variables 1`] = `
Object {
  "input": Object {
    "nested": Object {
      "nestedInput": "NESTEDINPUT",
    },
    "topInput": "TOPINPUT",
  },
}
`;

exports[`works on the nested structure: Output 1`] = `
Object {
  "__typename": "Query",
  "nested": Object {
    "name": "A",
  },
}
`;

exports[`works on the nestedFragment structure: Output 1`] = `
Object {
  "__typename": "Query",
  "listNested": Array [
    Object {
      "deeplyNested": Object {
        "name": "SECONDLEVEL",
      },
      "name": "FIRSTLEVEL",
    },
  ],
}
`;

exports[`works on the nestedNullable structure: Output 1`] = `
Object {
  "__typename": "Query",
  "nestedNullable": null,
}
`;

exports[`works on the repeatedFragment structure: Output 1`] = `
Object {
  "__typename": "Query",
  "first": Object {
    "name": "A",
  },
  "second": Object {
    "name": "A",
  },
}
`;

exports[`works on the simple structure: Output 1`] = `
Object {
  "__typename": "Query",
  "simple": "A",
}
`;
